subprojects/core/src/main/java/org/gradle/api/internal/file/archive/ZipFileTree.java

```diff
@@ -78,17 +82,17 @@ public DirectoryFileTree getMirror() {
 
     @Override
     public void visit(FileVisitor visitor) {
-        decompressionCache.useCache(() -> {
-            File zipFile = fileProvider.get();
-            if (!zipFile.exists()) {
-                throw new InvalidUserDataException(format("Cannot expand %s as it does not exist.", getDisplayName()));
-            }
-            if (!zipFile.isFile()) {
-                throw new InvalidUserDataException(format("Cannot expand %s as it is not a file.", getDisplayName()));
-            }
+        File zipFile = fileProvider.get();
+        if (!zipFile.exists()) {
+            throw new InvalidUserDataException(format("Cannot expand %s as it does not exist.", getDisplayName()));
+        }
+        if (!zipFile.isFile()) {
+            throw new InvalidUserDataException(format("Cannot expand %s as it is not a file.", getDisplayName()));
+        }
 
+        File expandedDir = getExpandedDir();
+        decompressionCache.useCache(expandedDir, () -> {
```

Comment:
I don't think so, but to be clear...

Are you worried about the `fileProvider` being unsafe concurrently, the contents of archive file changing concurrently or the expanded dir changing concurrently?

These checks are on the archive file itself and not the expanded directory (or the files inside it). I don't think there's a change in behavior if you have a task that extracts the zip and another one that deletes the expandedDir without holding the lock. 

We still only allow one thread per process to work inside a single expandedDir at a time.

Reply:
I was worried about seeing multiple exceptions instead of just one in an edge case; not about the safety of the visiting itself.  If we delay the existence and isFile checks until getting the cache lock, it helps avoid these, at the cost of moving a little more into the critical locking path.

It seems that seeing both the "does not exist" and "is not a file" exceptions for the same file (from different threads) is possible now, and that would be _really_ confusing.

Reply:
I was think just use some sort of "NO_FILE" sentinel value to get the `expandedDir` to give to the `useCache` call and get then emit the exceptions within the locked section.

But thinking more about your first example here, I don't think it's necessary.  Let's just leave this as is.